### PR TITLE
Features/54 wq mgmt plan pages

### PR DESCRIPTION
**Pull request by sraleci**
__Introducing...__
... Water Quality Management Plan!!

+ Added new Entities to Stormwater Tools:
  + Water Quality Management Plan (WQMP)
  + Water Quality Management Plan Document
+ Added CRUD for each
+ Added a bunch of lookup tables to use as attributes on plans
  + Priority
  + Status
  + Land Use
  + Development Type
+ Added Page Type for new WQMP Index
+ Added Stormwater Entity lookup for WQMP
+ Added Field Definition to make verbiage multi-tenant safe
+ Treatment BMP association to WQMP managed through:
  + Many-to-one on Treatment BMP Edit
  + One-to-many angular editor accessible from WQMP detail page

### CODE REVIEW

File: Source/Neptune.Web/Controllers/TreatmentBMPController.cs [hunk 1]
```diff
@@ -220,11 +217,11 @@ public PartialViewResult SummaryForMap(TreatmentBMPPrimaryKey treatmentBMPPrimar
         public JsonResult FindByName(string term)
         {
             var searchString = term.Trim();
-            var allTreatmentBMPsMatchingSearchString =
+            var allTreatmentBmPsMatchingSearchString =
```
Comment (NickLacunha) [inline Source/Neptune.Web/Controllers/TreatmentBMPController.cs]: "BmPs" makes me sad :(

File: Source/Neptune.Web/Models/WaterQualityManagementPlan.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+using Neptune.Web.Common;
+using Neptune.Web.Controllers;
+
+namespace Neptune.Web.Models
+{
+    public partial class WaterQualityManagementPlan : IAuditableEntity
+    {
+        public string AuditDescriptionString => $"Water Quality Management Plan \"{WaterQualityManagementPlanName}\"";
+
+        public string GetDetailUrl() => SitkaRoute<WaterQualityManagementPlanController>.BuildUrlFromExpression(c =>
```
Comment (NickLacunha) [inline Source/Neptune.Web/Models/WaterQualityManagementPlan.cs]: I'm used to seeing the Get*Url() methods as extension methods (e.g. in a static WaterQualityManagementPlanModelExtensions class), but I think I like this better.

File: Source/Neptune.Web/Controllers/TreatmentBMPController.cs [hunk 3]
```diff
@@ -152,20 +154,20 @@ private ViewResult ViewEdit(EditViewModel viewModel)
...
             var organizations = HttpRequestStorage.DatabaseEntities.Organizations.ToList();
+            var waterQualityManagementPlans = HttpRequestStorage.DatabaseEntities.WaterQualityManagementPlans.Where(x => x.StormwaterJurisdictionID == treatmentBMP.StormwaterJurisdictionID).ToList();
 
             if (ModelObjectHelpers.IsRealPrimaryKeyValue(viewModel.StormwaterJurisdictionID))
             {
                 var currentJurisdiction =
                     HttpRequestStorage.DatabaseEntities.StormwaterJurisdictions.GetStormwaterJurisdiction(
-                        viewModel.StormwaterJurisdictionID);
-
+                        viewModel.StormwaterJurisdictionID ?? ModelObjectHelpers.NotYetAssignedID);
```
[the remark this answers is not in the document]
Comment (sraleci) [reply Source/Neptune.Web/Controllers/TreatmentBMPController.cs]: Yeah, but I want to make appease the inspection gods.

File: Source/Neptune.Web/Controllers/TreatmentBMPController.cs [hunk 4]
```diff
@@ -115,16 +116,17 @@ public ActionResult New(EditViewModel viewModel)
...
         }
 
         [HttpGet]
         [TreatmentBMPEditFeature]
         public ViewResult Edit(TreatmentBMPPrimaryKey treatmentBMPPrimaryKey)
         {
             var treatmentBMP = treatmentBMPPrimaryKey.EntityObject;
-            var treatmentBMPLocationPoint = treatmentBMP.LocationPoint;
-            var viewModel = new EditViewModel(treatmentBMP, treatmentBMPLocationPoint);
+            var viewModel = new EditViewModel(treatmentBMP);
```
[the remark this answers is not in the document]
Comment (sraleci) [reply Source/Neptune.Web/Controllers/TreatmentBMPController.cs]: Seems that way, there was a lot of code the inspector was complaining wasn't needed anymore - especially in the Treatment BMP Controller - so I gutted it.

File: Source/Neptune.Web/Controllers/TreatmentBMPController.cs [hunk 5]
```diff
@@ -62,12 +63,12 @@ public ViewResult Index()
         [NeptuneViewFeature]
         public GridJsonNetJObjectResult<TreatmentBMP> TreatmentBMPGridJsonData()
         {
-            var treatmentBMPs = GetTreatmentBMPsAndGridSpec(out var gridSpec, CurrentPerson);
-            var gridJsonNetJObjectResult = new GridJsonNetJObjectResult<TreatmentBMP>(treatmentBMPs, gridSpec);
+            var treatmentBmps = GetTreatmentBmpsAndGridSpec(out var gridSpec, CurrentPerson);
```
[the remark this answers is not in the document]
Comment (sraleci) [reply Source/Neptune.Web/Controllers/TreatmentBMPController.cs]: I guess but I haaate it.

File: Source/Neptune.Web/Views/Shared/SiteLayout.cshtml [hunk 6]
```diff
@@ -22,7 +22,7 @@ Source code is available upon request via <support@sitkatech.com>.
 @inherits Neptune.Web.Views.Shared.SiteLayout
 
 @{
-    SetLayout("~/Views/Shared/NavAndHeaderLayout.cshtml", ViewDataTyped);
+    @SetLayout("~/Views/Shared/NavAndHeaderLayout.cshtml", ViewDataTyped)
```
Comment (NickLacunha) [inline Source/Neptune.Web/Views/Shared/SiteLayout.cshtml]: I'm not sure I understand what the difference between these is?

File: Source/Neptune.Web/Views/TreatmentBMP/EditViewModel.cs [hunk 7]
```diff
@@ -127,20 +125,17 @@ public void UpdateModel(Models.TreatmentBMP treatmentBMP, Person currentPerson)
...
-            var validationResults = new List<ValidationResult>();
-
-            var treatmentBMPsWithSameName = HttpRequestStorage.DatabaseEntities.TreatmentBMPs.Where(x => x.TreatmentBMPName == TreatmentBMPName);
-
-            if (treatmentBMPsWithSameName.Any(x => x.TreatmentBMPID != TreatmentBMPID))
+            if (HttpRequestStorage.DatabaseEntities.TreatmentBMPs.Where(x => x.TreatmentBMPName == TreatmentBMPName)
+                .Any(x => x.TreatmentBMPID != TreatmentBMPID))
             {
-                validationResults.Add(new SitkaValidationResult<EditViewModel, string>("A BMP with this name already exists.", x => x.TreatmentBMPName));
+                yield return new SitkaValidationResult<EditViewModel, string>("A BMP with this name already exists.",
```
Comment (NickLacunha) [inline Source/Neptune.Web/Views/TreatmentBMP/EditViewModel.cs]: yassssss

File: Source/Neptune.Web/Controllers/TreatmentBMPController.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (NickLacunha) [reply Source/Neptune.Web/Controllers/TreatmentBMPController.cs]: :+1:

File: Source/Neptune.Web/Controllers/TreatmentBMPController.cs (hunk 1 above)
Comment (sraleci) [reply Source/Neptune.Web/Controllers/TreatmentBMPController.cs]: Oh yeah, good catch. That's egregious.

File: Source/Neptune.Web/Models/WaterQualityManagementPlan.cs (hunk 2 above)
Comment (sraleci) [reply Source/Neptune.Web/Models/WaterQualityManagementPlan.cs]: Either one works and static might be a more performant option. I'll revisit this and probably use a static template, but I probably wont move it to the Model Extension.

File: Source/Neptune.Web/Views/Shared/SiteLayout.cshtml (hunk 6 above)
Comment (sraleci) [reply Source/Neptune.Web/Views/Shared/SiteLayout.cshtml]: Me either but I think the inspector prefers this one.

File: Source/Neptune.Web/Views/TreatmentBMP/EditViewModel.cs (hunk 7 above)
Comment (sraleci) [reply Source/Neptune.Web/Views/TreatmentBMP/EditViewModel.cs]: yaaaaaassss

Comment (sraleci) [reply Source/Neptune.Web/Views/TreatmentBMP/EditViewModel.cs]: Yas

**Pull request merged**